### DISCUSSION TITLE
FastAPI server-side HTTP sessions management libraries

### DISCUSSION THREAD

**Discussion by behai-nguyen** [fastapi/fastapi#11510] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
This question is not code related. I don't have a problem.
```

### Description

Please help me with the following query.

I have only learned FastAPI during this past three weeks.  

I have done a bit Flask before. With Flask, we can use the [Flask-Session](https://flask-session.readthedocs.io/en/latest/) package to manage server-side HTTP sessions: we can write our temporary session data into a MySQL table.

I have been looking for an equivalence package in FastAPI, I have not been able to find one.

Would you be able to suggest something, please?

Thank you and best regards.

### Operating System

Other

### Operating System Details

This question is not related to any operating system.

### FastAPI Version

0.110.2

### Pydantic Version

2.7.1

### Python Version

Python 3.10.1

### Additional Context

This is the [first FastAPI so-called web-server application](https://github.com/behai-nguyen/bh_database/tree/main/examples/fastapir) which I have written.

**Comment by hasansezertasan** [comment, marked as answer]
There is a library called [starsessions](https://github.com/alex-oleshkevich/starsessions), which could be what you need.

**Reply by behai-nguyen** [reply]
Thank you @hasansezertasan,

It looks promising. I will try it out. 

Thank you and best regards,

...behai.

**Comment by behai-nguyen** [comment]
Hi @hasansezertasan,

I have successfully used the ``starsession`` library in my application using Redis store. My application works exactly as programmed. I am happy. However, my integration test fails to access the Redis session store to remove the session data after the test finished. I have posted in the ``startsession`` discussion board [pytest can't seem to access Redis session store](https://github.com/alex-oleshkevich/starsessions/discussions/72). 

Have you come across this before, please? If you have, would be able to point out what I have missed, please? I am reproducing the post below. Thank you.

## pytest can't seem to access Redis session store

Hi,

Thank you for this library. I'm learning FastAPI and I find this library very helpful.

I have used its Redis session store without any problem in my application proper. However, my integration test using ``pytest`` does not seem to be able to "access" the Redis session store.

What have I missed please?

Content of ``/fastapi_learning/tests/conftest.py``:

```python
import pytest

from fastapi.testclient import TestClient

from . import test_main

@pytest.fixture(scope='module')
def app():
    return test_main.app

@pytest.fixture(scope='module')
def test_client(app):
    return TestClient(app)
```

Following is the ``/auth/logout`` endpoint which I am writing test for:

```python
def __login_page_context(inc_not_login_msg=False):
    context = {"title": "FastAPI Login"}
    if inc_not_login_msg:
        context.update({"message": "This session has not been logged in before"})

    return context

@router.post("/logout", response_class=HTMLResponse)
async def logout(request: Request):
    context = __login_page_context(
        inc_not_login_msg=request.session.get("access_token")==None)
    
    request.session.clear()
    
    return templates.TemplateResponse(request=request, name="auth/login.html", context=context)
```

When the test gets to this endpoint, ``request.session`` is ``{}``.

The integration test is not able to clear the session content from Redis. While the application is able to. That is, the application works as expected.

The call below has no effect during test, the corresponding Redis session does not get cleared:

```python
    request.session.clear()
```	

And this is my test method:

```python
@pytest.mark.auth_integration
def test_integration_valid_login(test_client):
    """
    Test /auth/token path with a valid credential.
    """

    try:
        login_data = {
            "username": "behai_nguyen@hotmail.com",
            "password": "password"
        }
        response = test_client.post(  
            "/auth/token", 
            data=login_data,
            headers={ 'Content-Type': 'application/x-www-form-urlencoded'}
        )

        assert response != None
        assert response.status_code == HTTPStatus.OK.value

        status = response.json()
        assert status['access_token'] == 'behai_nguyen@hotmail.com'
        assert status['token_type'] == 'bearer'

        print("\ncookie session -> ", test_client.cookies.get('session'))

    finally:
        # Logout. Clean up server sessions.
        test_client.post('/auth/logout')
```

``test_client.cookies.get('session')`` does have a valid session Id, which matches
the one in Redis, for example ``e87901248334d8f8e8d2b3fb4175b8c4``.

And ``test_client.post('/auth/logout')`` does go to the endpoint code.

Thank you and best regards,

...behai.

**Comment by behai-nguyen** [comment]
With the helps of ChatGPT, I have the solution for it:

```python
@pytest.mark.behai_only
def test_integration_valid_login(test_client):
    """
    Test /auth/token path with a valid credential.
    """

    try:
        login_data = {
            'username': 'behai_nguyen@hotmail.com',
            'password': 'password'
        }
        response = test_client.post(  
            '/auth/token', 
            data=login_data,
            headers={'Content-Type': 'application/x-www-form-urlencoded'}
        )

        assert response != None
        assert response.status_code == HTTPStatus.OK.value

        status = response.json()
        assert status['access_token'] == 'behai_nguyen@hotmail.com'
        assert status['token_type'] == 'bearer'

        session_cookie = response.cookies.get('session')

    finally:
        test_client.cookies = {'session': session_cookie}
        test_client.post('/auth/logout')
```
